### PR TITLE
[Utility] Local Proof of Stake CLI - Issue #112

### LINKED ISSUE #112: [Utility] Local Proof of Stake CLI
# Objective
Create an entry point to interact with the local proof of stake blockchain actors for testing, development and automation purposes.

# Origin Document
In order to test, develop and drive consensus in V1, we developed the [Pocket v1 client](https://github.com/pokt-network/pocket/blob/main/app/client/main.go) to send a few simple transactions such as printing node state, triggering next view , etc...

Similar to Tendermint's [abci-cli](https://docs.tendermint.com/master/app-dev/abci-cli.html#) tool, we need to expand the functionality of the pocket-dev-cli to fit our needs for the time being.

# Goals / Deliverables
- [x] Determine if the existing client suffices to fit these needs (i.e. passing params for addresses and values)
- [x] Define and document the list of commands to be added (e.g. stake, unstake, pause, unpause, send, etc...)
- [x] Update the node so the request is propagated appropriately
- [x] If needed, create the appropriate interfaces in the utility module (or other) to pass through or mock the request
Bonus:
- [x] Start rough notes for an RPC spec
- [x] Compile a listen of examples and references of other RPC specs for next steps

## General issue checklist
- [ ] CLI: Update the appropriate CHANGELOG (tracked in #193) 
- [ ] CLI: Update the README (tracked in #193)
- [x] RPC: Update the appropriate CHANGELOG
- [x] RPC: Update the README 
- [x] Node: Update the appropriate CHANGELOG
- [x] Node: Update the README 
- [x] Update the source code tree explanation
- [x] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)
- [x] Update any relevant global documentation & references
- [x] Document small issues / TODOs along the way

# Non-goals
- Create a fully-fledged RPC specification similar to [Tendermint](https://docs.tendermint.com/master/spec/rpc/#)
- Have functional end-to-end utility messages being submitted

# Testing Methodology
- **Utility specific tests**: `make test_utility_*` 
- **All tests**: `make test_all`
- **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)
Remove

Creator: @Olshansk 
Co-Owners: @andrewnguyen22

### PR DESCRIPTION
**Pull request by deblasis**
## Description
This PR aims at introducing CLI commands relative to the Utility module: 

- Send
- Stake
- EditStake
- Unstake
- Unpause
- ChangeParameter

Consequently, it introduces an RPC server (HTTP) and the ability to "point" the CLI at specific nodes via flags 

Fixes [issue/112](https://github.com/pokt-network/pocket/issues/112)

## Type of change
Please mark the options that are relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- [x] `make test_all`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md)

## Checklist
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] If applicable, I have made corresponding changes to related or global README
- [x] If applicable, I have added added new diagrams using [mermaid.js](https://mermaid-js.github.io)
- [x] If applicable, I have added tests that prove my fix is effective or that my feature works

## Demo

In order to test the features introduced, we need to start LocalNet

```bash
make compose_and_watch
```
image

in a separate console window, we can execute the newly introduced command

```bash
make swagger_ui
```
image

navigating to http://localhost:8080 we'll access a dockerized `swagger-ui` environment with injected our V1 RPC spec under development.

Please note that the openapi spec generates the RPC scaffolding and also the web client that's used by the CLI internally (new feature as well)

image

in here we can easily invoke the methods since we enabled CORS, that is disabled by default for security reasons. We can override this via a feature flag via environment variable, another feature added with this PR

image

Now we can simply test our RPC layer:

image
image

In the meantime the node is logging the corresponding requests:
image

How about the CLI?

Since we haven't -yet- packaged it into a binary we can run it from the source-code but soon we'll simply replace the `go run` with a `pokt-client` or `pocketcli` or any name @Olshansk can come up with given his track record :)

```bash
go run app/client/*.go
```
image

This could look daunting, but our users are generally speaking quite tech savvy and used to this popular format (offered by the library `cobra`).

Documentation is also generated in markdown starting having the CLI help as a source. (new feature) [app/client/cli/doc/README.md](https://github.com/deblasis/pocket/blob/bd9f0dd7ee31a523c7e9db1b3d2a03708f391d90/app/client/cli/doc/README.md#L9)
image

Now to the commands, there are some simple ones you might have spotted from the RPC:
image
image

We can obviously target a specific host as well
image

The tooling makes it easy to add new commands starting from the spec, these ones were implemented in few minutes:
image

I was inspired by the recent chain halt in V0

image

image

The 🥩ier stuff is about stake, unstake and these other commands.

As required in the initial issue, this PR limits itself to submitting the respective payload and make sure that signing and verification server side works properly.

Since we don't have a keybase yet, we need to supply a private key json that by default resides in the root folder (where you run `make` commands basically).

A test `pk.json` could be:
```
"e7760141c2672178b28360a8cf80ff3a9d5fd579990317b9afcb2091426ffe75dc12b26584c057be33fcc8e891a483250581e38fe2bc9d62c1a1341c5e85b667"
```
(note the double quotes... it's JSON! :) )

if we test a command 
image

(please forgive the non-realistic arguments)
image

The CLI packages the message, signs it using the private key and sends it to the RPC, which calls `CheckTransaction` and if there are no errors we return HTTP 200.

Now, if we try for a second time any other command that interacts with the Utility context we get this

image

But that's something unrelated that we'll fix separately.

fin

### CODE REVIEW

File: shared/config/defaults.go [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package config
+
+const (
```
Comment (Olshansk) [inline shared/config/defaults.go:3]: NOTE: Will need to be refactored after merging with https://github.com/pokt-network/pocket/pull/166

File: shared/config/config.go [hunk 2]
```diff
@@ -15,6 +15,11 @@ type Config struct {
 	RootDir       string                 `json:"root_dir"`
 	GenesisSource *genesis.GenesisSource `json:"genesis_source"` // TECHDEBT(olshansky): we should be able to pass the struct in here.
 
+	// DISCUSS(deblasis): should we scope this into, let's say, RPCConfig?
```
Comment (Olshansk) [inline shared/config/config.go:18]: 💯 yes

File: shared/config/config.go [hunk 3]
```diff
@@ -15,6 +15,11 @@ type Config struct {
...
+	RPCPort      string `json:"rpc_port"`       // TODO(deblasis): still unused, need to define default value, "8081" ?
```
Comment (Olshansk) [inline shared/config/config.go:19]: Let's stay consistent with tendermint for ports: https://docs.cosmos.network/master/core/grpc_rest.html

It'll be something people are used to.

File: app/pocket/main.go [hunk 4]
```diff
@@ -22,6 +23,12 @@ func main() {
 	}
 
 	cfg := config.LoadConfig(*config_filename)
+	// DISCUSS(deblasis): poor man's feature flag: no port => no RPC
+	if cfg.RPCPort != "" {
+		go rpc.StartRPC(cfg.RPCPort, cfg.RPCTimeout)
+	} else {
+		log.Println("RPC server: OFFLINE")
```
Comment (Olshansk) [inline app/pocket/main.go:30]: s/RPC server/[WARN] RPC server

File: app/pocket/main.go [hunk 5]
```diff
@@ -31,4 +38,5 @@ func main() {
 	if err = pocketNode.Start(); err != nil {
 		log.Fatalf("Failed to start pocket node: %s", err)
 	}
+
```
Comment (Olshansk) [inline app/pocket/main.go:41]: rm

File: app/pocket/main.go [hunk 6]
```diff
@@ -22,6 +23,12 @@ func main() {
 	}
 
 	cfg := config.LoadConfig(*config_filename)
+	// DISCUSS(deblasis): poor man's feature flag: no port => no RPC
```
Comment (Olshansk) [inline app/pocket/main.go:26]: Maybe adding an `EnableRPC` in the `RPCConfig` we'll add?

File: app/client/rpc/server.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/julienschmidt/httprouter"
+)
+
+//TODO(deblasis): still unused, need to gather feedback on RPC routes below before implementing everything, for now I copied some code from v0 over as a blueprint for style
```
Comment (Olshansk) [inline app/client/rpc/server.go:15]: Let's split out the server and the cli into two PRs.

I'm personally always in favor of iterating on "non-functional" but small, clear and easy-to-approach code in main

File: app/client/cli/account.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+func accountCommands() (cmds []*cobra.Command) {
+
+	sendCmd := &cobra.Command{
+		Use:     "Send <from> <to> <amount>",
+		Short:   "Send <from> <to> <amount>",
+		Long:    "Sends <amount> to address <to> from address <from>",
+		Aliases: []string{"send"},
+		Args:    cobra.ExactArgs(3),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// TODO(deblasis): update when we have keybase
```
Comment (Olshansk) [inline app/client/cli/account.go:37]: For keybase related followups, you can even do `TODO(pocket/issues/150)`

File: app/client/cli/cmd.go [hunk 9]
```diff
@@ -0,0 +1,36 @@
+package cli
+
+import (
+	"context"
+
+	"github.com/pokt-network/pocket/app"
+	"github.com/spf13/cobra"
+)
+
+const CLIExecutableName = "client"
```
Comment (Olshansk) [inline app/client/cli/cmd.go:10]: lowerCamelCase
Screen Shot 2022-08-23 at 7 12 43 PM


https://stackoverflow.com/a/22688926/768439

File: app/client/cli/actor.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+			if err != nil {
+				return err
+			}
+			// currently ignored since we are using the address from the PrivateKey
+			// fromAddr := crypto.AddressFromString(args[0])
+			amount := args[1]
+
+			_ = &types.MessageStake{
+				PublicKey:     pk.PublicKey().Bytes(),
+				Chains:        []string{}, // TODO(deblasis): 👀
```
Comment (Olshansk) [inline app/client/cli/actor.go:63]: This is still in flux, but we will probably either end up having something similar to v0:

Screen Shot 2022-08-23 at 7 21 28 PM

Or be able to pass in a path to a config to make it easier:

Screen Shot 2022-08-23 at 7 18 56 PM

Lets leave as is for now and cc @andrewnguyen22 for context

File: app/client/cli/debug.go [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+		Short: "Debug utility for rapid development",
+		Args:  cobra.ExactArgs(0),
+		RunE:  runDebug,
+	}
+
+	return c
+}
+
+func runDebug(cmd *cobra.Command, args []string) error {
+
```
Comment (Olshansk) [inline app/client/cli/debug.go:46]: rm 😅

File: app/client/cli/debug.go [hunk 12]
```diff
@@ -0,0 +1,144 @@
+package cli
```
Comment (Olshansk) [inline app/client/cli/debug.go:1]: Not actionable: Big fan of how this is in a separate file now.

We can do something similar for `infra` soon. CC @okdas @vlad-she

File: app/client/cli/gov.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+			key := args[1]
+			value := args[2]
+
+			pbValue, err := anypb.New(wrapperspb.String(value))
+			if err != nil {
+				return err
+			}
+
+			_ = &types.MessageChangeParameter{
+				Signer:         []byte{}, // TODO(deblasis): 👀 same as owner?
```
Comment (Olshansk) [inline app/client/cli/gov.go:57]: Can you add a `TODO(andrew)` in `message.proto` to document the difference? 

For now, you can use the same ones.

The `owner` is the owner of the funds staked (where the rewards go) and the `signer` is the operator. Most of the time, if you run your own node, it's the same thing.

**Review by Olshansk — CHANGES_REQUESTED** [review]
tl;dr Thank you for this! Simply having this as a baseline to look at, update and edit will make the implementation 1000x easier 💯 

I took a first high-level round of reviews (see screenshot since incomplete) but wanted to leave some comments.

1. Let's split the server and the client into two different PRs
2. Left a couple actionables throughout
3. I think merging it in with all the TODOs (no keybase, etc) AFTER the merge with the major config changes is more than a sufficient first step for us to iterate on.

Screen Shot 2022-08-23 at 7 28 17 PM

File: app/app.go [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/pokt-network/pocket/shared"
+	"github.com/pokt-network/pocket/shared/config"
+	"github.com/pokt-network/pocket/shared/crypto"
+	"github.com/pokt-network/pocket/shared/modules"
+	"github.com/pokt-network/pocket/shared/telemetry"
+	"github.com/pokt-network/pocket/shared/types/genesis"
+)
+
+var (
+	Config       *config.Config
```
[the remark this answers is not in the document]
Comment (deblasis) [reply app/app.go]: addressed with a refactoring as I am figuring stuff out

File: app/app.go [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+	P2pMod       modules.P2PModule
+	ConsensusMod modules.ConsensusModule
+
+	modInitOnce sync.Once
+)
+
+func Init(remoteCLIURL string) {
```
[the remark this answers is not in the document]
Comment (deblasis) [reply app/app.go]: addressed with a refactoring as I am figuring stuff out

File: app/app.go [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+
+	modInitOnce.Do(func() {
+
```
[the remark this answers is not in the document]
Comment (deblasis) [reply app/app.go]: 🤦‍♂️

File: app/app.go [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+		pk, err := crypto.GeneratePrivateKey()
+		if err != nil {
+			log.Fatalf("[ERROR] Failed to generate private key: %v", err)
+		}
+
+		Config := &config.Config{
```
[the remark this answers is not in the document]
Comment (deblasis) [reply app/app.go]: Nice! Watching the PR 👀

File: app/app.go [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+			GenesisSource: &genesis.GenesisSource{
+				Source: &genesis.GenesisSource_File{
+					File: &genesis.GenesisFile{
+						Path: "build/config/genesis.json",
```
[the remark this answers is not in the document]
Comment (deblasis) [reply app/app.go]: moved into config `New` for now but watching #166 as well, as soon as it's merged I'll update everything

File: app/client/cli/account.go [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+		Args:    cobra.ExactArgs(0),
+	}
+
+	cmd.AddCommand(accountCommands()...)
+
+	return cmd
+}
+
+func accountCommands() (cmds []*cobra.Command) {
+
```
[the remark this answers is not in the document]
Comment (deblasis) [reply app/client/cli/account.go]: ARGGGG! 

OK, now I am testing out naively running [gofumpt](https://github.com/mvdan/gofumpt) before committing :
```
gofumpt -l -w <PATH>
```
having hooks or something that fixes this automatically would be great, I have subscribed to the issue #175 as well, tnx!

File: app/client/cli/account.go (hunk 8 above)
Comment (deblasis) [reply app/client/cli/account.go]: done

File: app/client/cli/cmd.go (hunk 9 above)
Comment (deblasis) [reply app/client/cli/cmd.go]: done

File: app/client/cli/debug.go (hunk 11 above)
Comment (deblasis) [reply app/client/cli/debug.go]: LOL

File: app/client/cli/gov.go (hunk 13 above)
Comment (deblasis) [reply app/client/cli/gov.go:56]: I don't think it's an AI for Andrew, your explanation confirms my hunch and it makes total sense.

These "TODO 👀" are a reminder for me to dig deeper before raising questions (for which comments like yours make my life easier tho). Normally I wouldn't share but since I felt this PR could be a bit of a blocker, I wanted to share as much as possible so that you guys can figure out where we are at with this, including my doubts hehe.

One thing I figured is that I think I have to construct these messages server side and not on the client for various reasons:
- implementation details leaks + tight coupling
- on the server side (node) we know the signer address so I can populate this value correctly I guess

File: app/client/rpc/server.go (hunk 7 above)
Comment (deblasis) [reply app/client/rpc/server.go]: Agreed, I will address the current AIs and request review again when I have 2 separate PRs.

File: app/pocket/main.go (hunk 6 above)
Comment (deblasis) [reply app/pocket/main.go]: Done

File: app/pocket/main.go (hunk 4 above)
Comment (deblasis) [reply app/pocket/main.go]: Done

File: app/pocket/main.go (hunk 5 above)
Comment (deblasis) [reply app/pocket/main.go]: <ackward silence>

LOL

:)

File: p2p/module.go [hunk 20]
```diff
@@ -88,6 +88,7 @@ func (m *p2pModule) Start() error {
 	}
 	m.network.SetBus(m.GetBus())
 
+	// DISCUSS(deblasis): this seems redundant (see line above)
```
[the remark this answers is not in the document]
Comment (deblasis) [reply p2p/module.go]: Done

File: shared/config/config.go (hunk 2 above)
Comment (deblasis) [reply shared/config/config.go]: Done

File: shared/config/config.go (hunk 3 above)
Comment (deblasis) [reply shared/config/config.go]: Ok, I'll use port 26657

Probably out of scope for this PR(s) but now a question I wanted to ask earlier too:

Should we offer GRPC + GRPC-Gateway APIs at some point in a Cosmos-like fashion?

I have done that in the past and as you can probably tell, I am a fan of protobufs and GRPC :)

**Comment by andrewnguyen22** [comment]
I'm assuming this is succeeded by #176 and #177?

If so - let's close this

**Comment by Olshansk** [comment]
@deblasis Per what Andrew said. Can we close this given that it was split out in the other 2 PRs?

**Comment by deblasis** [comment]
> EDIT: Nvm, I see comments like these now:

Yep, this PR basically contains the commits that represent the "glue" between the other two PRs.

**Comment by Olshansk** [comment]
@deblasis I've reviewed [1/2] and will take a look at [2/2], but does this PR need to be reviewed as well?

**Comment by deblasis** [comment]
At this point it would be just unnecessary/repetitive since it includes the other two PRs and just some extra code that's not really easy to isolate.

When the other two are merged we'll see only the extra bits (that we can review) and we can also use this PR to track/review the remaining work that requires both CLI and RPC. For instance a quick gif/curl demo or something.

**Comment by Olshansk** [comment]
Gotta call the 🚒 because that demo is frikkin 🔥 🔥 

Going to go through all the steps mysef when we're closer to merging it in, but otherwise, this is **exactly** what's going to get us to the next milestone demo.

Once it's in, can I make a request that you demo it to the team in protocol hour? :)

> Since we haven't -yet- packaged it into a binary we can run it from the source-code but soon we'll simply replace the go run with a pokt-client or pocketcli or any name @Olshansk can come up with given his track record :)

I'm going to need to meditate on this one since it'll be around for the ages. The goal is to have all the chains and tools in your `pocket`, so it might very well be that simple.

**Comment by deblasis** [comment]
> Gotta call the 🚒 because that demo is frikkin 🔥 🔥

😂

> Going to go through all the steps mysef when we're closer to merging it in, but otherwise, this is **exactly** what's going to get us to the next milestone demo.
> 
> Once it's in, can I make a request that you demo it to the team in protocol hour? :)

Of course, very happy to do that. We need feedback, particularly from folks that interact a lot with V0 and other chains so that we can adjust early on.

> > Since we haven't -yet- packaged it into a binary we can run it from the source-code but soon we'll simply replace the go run with a pokt-client or pocketcli or any name @Olshansk can come up with given his track record :)
> 
> I'm going to need to meditate on this one since it'll be around for the ages. The goal is to have all the chains and tools in your `pocket`, so it might very well be that simple.

Makes sense but let's remember that we also have the node binary and we might want to differentiate them. There are also projects that have the same entrypoint for CLI and application executable (Hashicorp tools for example, which I know a bit).

Food for thought

**Comment by Olshansk** [comment]
@deblasis Is it correct to change the base of this PR to `176` until that one is merged in? Just trying to make it easier to only review the delta.

Screen Shot 2022-11-07 at 10 52 41 AM

**Comment by deblasis** [comment]
> @deblasis Is it correct to change the base of this PR to `176` until that one is merged in? Just trying to make it easier to only review the delta.
> 
> Screen Shot 2022-11-07 at 10 52 41 AM

Done!

Since the branches were created off my private repo, I had to create a `pock-network/issue/utility_local_proof_of_stake_cli_RPC` branch and change the base of this PR to that. I have never done that before but it seems to work.

File: app/client/cli/consensus.go [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+
+	cmd.AddCommand(consensusCommands()...)
+
+	return cmd
+}
+
+func consensusCommands() []*cobra.Command {
+	cmds := []*cobra.Command{
+		{
+			Use:     "RoundState",
```
Comment (Olshansk) [inline app/client/cli/consensus.go:31]: Inspired by cosmos ;)

File: app/client/cli/consensus.go [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+
+				fmt.Printf("%d/%d/%d\n", response.JSONDefault.Height, response.JSONDefault.Round, response.JSONDefault.Step)
+
+				return nil
+			},
+		},
+		{
+			Use:     "Height",
+			Short:   "Returns the Height",
+			Long:    "Height returns the height in the consensus state",
```
Comment (Olshansk) [inline app/client/cli/consensus.go:49]: NIT:

```suggestion
			Long:    "Height returns the height in the node's current consensus state",
```

Ditto below.

File: app/client/cli/consensus.go [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+	}
+	return cmds
+}
+
+func getConsensusRoundState(cmd *cobra.Command) (*rpc.GetV1ConsensusRoundStateResponse, error) {
+	client, err := rpc.NewClientWithResponses(remoteCLIURL)
+	if err != nil {
+		return nil, nil
+	}
+	response, err := client.GetV1ConsensusRoundStateWithResponse(cmd.Context())
```
Comment (Olshansk) [inline app/client/cli/consensus.go:103]: `s/GetV1ConsensusRoundStateWithResponse/GetV1ConsensusState`
OR
`s/GetV1ConsensusRoundStateWithResponse/GetV1ConsensusStateWithResponse`

File: app/client/cli/doc/CHANGELOG.md [hunk 24]
```diff
@@ -13,6 +13,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Fixed message signing
 - Reporting RPC StatusCode and body
+- System commands working end-to-end
+- Added Consensus RoundState commands
```
Comment (Olshansk) [inline app/client/cli/doc/CHANGELOG.md:17]: Let's just call them `State` commands, and `RoundState` is an implied "subcommand". The terminology is more future-proof that way.

File: app/client/cli/system.go [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+
+func systemCommands() []*cobra.Command {
+	cmds := []*cobra.Command{
+		{
+			Use:     "Health",
+			Short:   "RPC endpoint liveness",
+			Long:    "Performs a simple liveness check on the node RPC endpoint",
+			Aliases: []string{"health"},
+			RunE: func(cmd *cobra.Command, args []string) error {
+
```
Comment (Olshansk) [inline app/client/cli/system.go:37]: NIT: remove newline. Ditto below.

File: app/client/cli/system.go [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+				if err != nil {
+					return nil
+				}
+				response, err := client.GetV1HealthWithResponse(cmd.Context())
+				if err != nil {
+					return unableToConnectToRpc(err)
+				}
+				statusCode := response.StatusCode()
+				if statusCode == http.StatusOK {
+					fmt.Printf("✅ RPC reporting healthy status for node @ \033[1m%s\033[0m\n\n%s", remoteCLIURL, response.Body)
```
Comment (Olshansk) [inline app/client/cli/system.go:48]: Is the `\033[1m%` in place for color? If so, can you add a tiny comment saying `URL is color X`

File: app/client/cli/utils.go [hunk 27]
```diff
@@ -190,3 +190,13 @@ func attachPwdFlagToSubcommands() []cmdOption {
 		c.Flags().StringVar(&pwd, "pwd", "", "passphrase used by the cmd, non empty usage bypass interactive prompt")
 	}}
 }
+
+func unableToConnectToRpc(err error) error {
+	fmt.Printf("❌ Unable to connect to the RPC @ \033[1m%s\033[0m\n\nError: %s", remoteCLIURL, err)
```
Comment (Olshansk) [inline app/client/cli/utils.go:195]: Good use of emoji 🥇

File: app/client/cli/utils.go [hunk 28]
```diff
@@ -190,3 +190,13 @@ func attachPwdFlagToSubcommands() []cmdOption {
...
+	return nil
+}
+
+func rpcResponseCodeUnhealthy(statusCode int, response []byte) error {
+	fmt.Printf("❌ RPC reporting unhealthy status HTTP %d @ \033[1m%s\033[0m\n\n%s", statusCode, remoteCLIURL, response)
```
Comment (Olshansk) [inline app/client/cli/utils.go:200]: This might be overkill so consider it optional

```go
func formatText(s string) string {
   return  fmt.Printf("\033[1m%s\033[0m", s)
}
```

File: build/deployments/docker-compose.yaml [hunk 29]
```diff
@@ -46,8 +46,9 @@ services:
     # Needed for DLV debugging
     security_opt:
       - "seccomp:unconfined"
+    environment:
+      - POCKET_RPC_USE_CORS=true
```
Comment (Olshansk) [inline build/deployments/docker-compose.yaml:50]: Were there issues with CORS disabled in this commit? Just curious to understand

**Review by Olshansk — CHANGES_REQUESTED** [review]
Just a few minor comments bust mostly ready.

File: rpc/doc/README.md [hunk 30]
```diff
@@ -1,7 +1,20 @@
-# RPC
+# RPC <!-- omit in toc -->
 
 This document is meant to be a starting point/placeholder for a full-fledged RPC specification that allows interaction with the nodes.
 
+## Contents <!-- omit in toc -->
```
[the remark this answers is not in the document]
Comment (deblasis) [reply rpc/doc/README.md]: Yep, nice one

File: build/deployments/docker-compose.yaml (hunk 29 above)
Comment (deblasis) [reply build/deployments/docker-compose.yaml:50]: No, by default I left CORS disabled (I'd rather have it opt-in). This environment variable simply enables it.

File: app/client/cli/utils.go (hunk 27 above)
Comment (deblasis) [reply app/client/cli/utils.go:195]: 🫡

File: app/client/cli/utils.go (hunk 28 above)
Comment (deblasis) [reply app/client/cli/utils.go]: If it's not clear like I understand from https://github.com/pokt-network/pocket/pull/169#discussion_r1016924253

it means that something like this:

https://github.com/pokt-network/pocket/pull/169/commits/f2c765fa18a3e0fd02134907db7f537dc1fc591e

is not overkill and it's better than a comment.

File: app/client/cli/system.go (hunk 26 above)
Comment (deblasis) [reply app/client/cli/system.go]: https://github.com/pokt-network/pocket/pull/169/commits/f2c765fa18a3e0fd02134907db7f537dc1fc591e

File: app/client/cli/doc/commands/client_Application_Stake.md [hunk 31]
```diff
@@ -1,10 +1,18 @@
 ## client Application Stake
 
-Stake an actor (Application) in the network.
+Stake a node in the network. Custodial stake uses the same address as operator/output for rewards/return of staked funds.
 
 ### Synopsis
 
-Stake the Application actor into the network, making it available for service.
+Stake the node into the network, making it available for service.
```
Comment (Olshansk) [inline app/client/cli/doc/commands/client_Application_Stake.md:7]: Optional nit: this formats as a single line in markdown github. Was that the intention or did you mean to split it out?

Screen Shot 2022-11-09 at 11 03 19 AM

I'd add new lines to make reading easier but doesn't really matter.

**Review by Olshansk — APPROVED** [review]

File: app/client/cli/doc/commands/client_Application_Stake.md (hunk 31 above)
Comment (deblasis) [reply app/client/cli/doc/commands/client_Application_Stake.md:7]: Makes sense, updated 👍

**Review by Olshansk — APPROVED** [review]
